### PR TITLE
add resnet sample for hptuning with cmle tpu

### PR DESCRIPTION
**Pull request by dizcology**
This PR adds two samples both based on the official ResNet on TPU sample, modified to be runnable on CMLE:

1. hyperparameter tuning with `tf.metric`

1. hyperparameter tuning with `cloudml-hypertune`

1. training

### CODE REVIEW

File: tpu/hptuning/resnet-hypertune/resnet/hypertune_hook.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    def __init__(self):
+        self.hypertune = hypertune.HyperTune()
+        self.hp_metric_tag = os.environ.get('CLOUD_ML_HP_METRIC_TAG', '')
+        self.trial_id = os.environ.get('CLOUD_ML_TRIAL_ID', 0)
+
+    def end(self, session):
+        step_variable = session.graph.get_collection('global_step')
+        global_step = session.run(step_variable)[0]
+
+        tf.logging.info('DEBUG: HypertuneHook called, tag: {}, trial_id: {}, global_step: {}'.format(self.hp_metric_tag, self.trial_id, global_step))
```
Comment (dizcology) [inline tpu/hptuning/resnet-hypertune/resnet/hypertune_hook.py]: remove DEBUG

File: tpu/hptuning/resnet-hypertune/config_resnet_hypertune.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+trainingInput:
+  scaleTier: CUSTOM
+  masterType: standard
+  workerCount: 1
+  workerType: cloud_tpu
+  hyperparameters:
+    goal: MAXIMIZE
+    enableTrialEarlyStopping: true
+    hyperparameterMetricTag: my_top_5_accuracy
+    maxTrials: 3
```
Comment (0olwzo0) [inline tpu/hptuning/resnet-hypertune/config_resnet_hypertune.yaml]: lets put more trials in the config. The single round run won't benefit a lot from hyperparameter tuning.

File: tpu/hptuning/resnet-hptuning/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    ```
+    gsutil mb gs://[YOUR_GCS_BUCKET]
+    export GCS_BUCKET="gs://[YOUR_GCS_BUCKET]"
+    ```
+
+1. Run the sample.  The included script will train ResNet-50 for 1024 steps using a fake dataset.
+
+    ```
+    cd cloudml-samples/tpu/hpruning/resnet-hptuning
+    bash submit.sh
```
Comment (0olwzo0) [inline tpu/hptuning/resnet-hptuning/README.md]: it should be submit_resnet.sh?

File: tpu/training/resnet/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    ```
+    gsutil mb gs://[YOUR_GCS_BUCKET]
+    export GCS_BUCKET="gs://[YOUR_GCS_BUCKET]"
+    ```
+
+1. Run the sample.  The included script will train ResNet-50 for 1024 steps using a fake dataset.
+
+    ```
+    cd cloudml-samples/tpu/training/resnet
+    bash submit.sh
```
Comment (0olwzo0) [inline tpu/training/resnet/README.md]: same here submit_resnet.sh

File: tpu/hptuning/resnet-hypertune/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    ```
+    gsutil mb gs://[YOUR_GCS_BUCKET]
+    export GCS_BUCKET="gs://[YOUR_GCS_BUCKET]"
+    ```
+
+1. Run the sample.  The included script will train ResNet-50 for 1024 steps using a fake dataset.
+
+    ```
+    cd cloudml-samples/tpu/hptuning/resnet-hypertune
+    bash submit.sh
```
Comment (0olwzo0) [inline tpu/hptuning/resnet-hypertune/README.md]: submit_resnet_hypertune.sh?

File: tpu/hptuning/resnet-hypertune/config_resnet_hypertune.yaml (hunk 2 above)
Comment (dizcology) [reply tpu/hptuning/resnet-hypertune/config_resnet_hypertune.yaml]: done.

File: tpu/hptuning/resnet-hptuning/README.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
+
+## Introduction
+
+This sample shows how to run hyperparameter tuning jobs on Cloud Machine Learning Engine with Cloud TPUs using TensorFlow's `tf.metrics`.
```
Comment (puneith) [inline tpu/hptuning/resnet-hptuning/README.md]: Not sure it's needed to mention "..using TF tf.metrics"

File: tpu/hptuning/resnet-hptuning/resnet/imagenet_input.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+from __future__ import print_function
+
+import os
+
+import tensorflow as tf
+
+from . import resnet_preprocessing
+
+
+def image_serving_input_fn():
```
Comment (puneith) [inline tpu/hptuning/resnet-hptuning/resnet/imagenet_input.py]: Why is this function not returning anything. Also why is there a need to have ServingInputReceiver in an inner function and not directly in serving?

File: tpu/hptuning/resnet-hptuning/resnet/resnet_main.py [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+from . import resnet_model
+from tensorflow.contrib import summary
+from tensorflow.contrib.tpu.python.tpu import bfloat16
+from tensorflow.contrib.tpu.python.tpu import tpu_config
+from tensorflow.contrib.tpu.python.tpu import tpu_estimator
+from tensorflow.contrib.tpu.python.tpu import tpu_optimizer
+from tensorflow.contrib.training.python.training import evaluation
+from tensorflow.python.estimator import estimator
+
+FLAGS = flags.FLAGS
```
Comment (puneith) [inline tpu/hptuning/resnet-hptuning/resnet/resnet_main.py]: change from flags => argsparse

File: tpu/hptuning/resnet-hptuning/resnet/resnet_main.py [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+  trial_id = os.environ.get('CLOUD_ML_TRIAL_ID', 0)
+  if trial_id:
+    if FLAGS.export_dir:
+      FLAGS.export_dir = os.path.join(
+        FLAGS.export_dir, 'trial_{}'.format(trial_id))
+    FLAGS.model_dir = os.path.join(
+      FLAGS.model_dir, 'trial_{}'.format(trial_id))
+
+  tpu_cluster_resolver = tf.contrib.cluster_resolver.TPUClusterResolver(
+      FLAGS.tpu,
```
Comment (puneith) [inline tpu/hptuning/resnet-hptuning/resnet/resnet_main.py]: Same FLAGS comment as above

File: tpu/hptuning/resnet-hptuning/resnet/resnet_main.py [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+def main(unused_argv):
+  trial_id = os.environ.get('CLOUD_ML_TRIAL_ID', 0)
+  if trial_id:
+    if FLAGS.export_dir:
+      FLAGS.export_dir = os.path.join(
+        FLAGS.export_dir, 'trial_{}'.format(trial_id))
+    FLAGS.model_dir = os.path.join(
+      FLAGS.model_dir, 'trial_{}'.format(trial_id))
+
+  tpu_cluster_resolver = tf.contrib.cluster_resolver.TPUClusterResolver(
```
Comment (puneith) [inline tpu/hptuning/resnet-hptuning/resnet/resnet_main.py]: This can benefit from explaining why TPUCLusterResolver is needed (as in difference in workflow compared to CPU/GPU). Or that will be part of blog?

File: tpu/hptuning/resnet-hptuning/resnet/resnet_main.py [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+      # The guide to serve a exported TensorFlow model is at:
+      #    https://www.tensorflow.org/serving/serving_basic
+      tf.logging.info('Starting to export model.')
+      resnet_classifier.export_savedmodel(
+          export_dir_base=FLAGS.export_dir,
+          serving_input_receiver_fn=imagenet_input.image_serving_input_fn)
+
+if __name__ == '__main__':
+  tf.logging.set_verbosity(tf.logging.INFO)
+  tf.app.run()
```
Comment (puneith) [inline tpu/hptuning/resnet-hptuning/resnet/resnet_main.py]: change tf.app.run() to simply python call as main(...). No need for this invocation.

File: tpu/hptuning/resnet-hptuning/resnet/resnet_preprocessing.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+  return image
+
+
+def _flip(image):
+  """Random horizontal image flip."""
+  image = tf.image.random_flip_left_right(image)
+  return image
+
+
+def preprocess_for_train(image_bytes):
```
Comment (puneith) [inline tpu/hptuning/resnet-hptuning/resnet/resnet_preprocessing.py]: why is train and eval pre-process different? Also there is only one line difference to not warrant duplicate.

File: tpu/hptuning/resnet-hypertune/resnet/hypertune_hook.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+        self.trial_id = os.environ.get('CLOUD_ML_TRIAL_ID', 0)
+
+    def end(self, session):
+        step_variable = session.graph.get_collection('global_step')
+        global_step = session.run(step_variable)[0]
+
+        tf.logging.info('HypertuneHook called, tag: {}, trial_id: {}, global_step: {}'.format(self.hp_metric_tag, self.trial_id, global_step))
+
+        # The name of the tensor is given in metric_fn in resnet_main_hypertune.py.
+        metric_tensor = session.graph.get_tensor_by_name('top_5_accuracy/value:0')
```
Comment (puneith) [inline tpu/hptuning/resnet-hypertune/resnet/hypertune_hook.py]: why getting metric tensor using string key?

File: tpu/hptuning/resnet-hypertune/resnet/imagenet_input.py [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+import os
+
+import tensorflow as tf
+
+from . import resnet_preprocessing
+
+
+def image_serving_input_fn():
+  """Serving input fn for raw images."""
+
```
Comment (puneith) [inline tpu/hptuning/resnet-hypertune/resnet/imagenet_input.py]: I think I saw duplicate code above. Serving function etc.

File: tpu/hptuning/resnet-hptuning/resnet/resnet_model.py [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+
+  inputs = conv2d_fixed_padding(
+      inputs=inputs, filters=filters, kernel_size=3, strides=1,
+      data_format=data_format)
+  inputs = batch_norm_relu(inputs, is_training, relu=False, init_zero=True,
+                           data_format=data_format)
+
+  return tf.nn.relu(inputs + shortcut)
+
+
```
Comment (puneith) [inline tpu/hptuning/resnet-hptuning/resnet/resnet_model.py]: why do we have duplicate files (code)?

File: tpu/hptuning/resnet-hptuning/resnet/resnet_preprocessing.py [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+
+IMAGE_SIZE = 224
+CROP_PADDING = 32
+MEAN_RGB = [0.485, 0.456, 0.406]
+STDDEV_RGB = [0.229, 0.224, 0.225]
+
+
+def distorted_bounding_box_crop(image_bytes,
+                                bbox,
+                                min_object_covered=0.1,
```
Comment (puneith) [inline tpu/hptuning/resnet-hptuning/resnet/resnet_preprocessing.py]: same duplicate files (code) comment.

File: tpu/hptuning/resnet-hptuning/README.md (hunk 6 above)
Comment (dizcology) [reply tpu/hptuning/resnet-hptuning/README.md]: This is to contrast the other possible ways of reporting metrics (`tf.summary`, and `cloudml.hypertune`).

File: tpu/hptuning/resnet-hptuning/resnet/imagenet_input.py (hunk 7 above)
Comment (dizcology) [reply tpu/hptuning/resnet-hptuning/resnet/imagenet_input.py]: The hptuning example is entirely based on the official ResNet TPU sample version r1.8 and I made only minimal changes required to use CMLE's hptuning service.  I would prefer to minimize diverging from the official (training) sample.

File: tpu/hptuning/resnet-hypertune/resnet/imagenet_input.py (hunk 14 above)
Comment (dizcology) [reply tpu/hptuning/resnet-hypertune/resnet/imagenet_input.py]: `tpu/hptuning/resnet-hypertune` and `tpu/hptuning/resnet-hptuning` are both based on the same official sample.  The goal is to show two different ways of using the hptuning service, while keeping both samples self-contained (as in, clone-runnable).

File: tpu/hptuning/resnet-hypertune/resnet/hypertune_hook.py (hunk 13 above)
Comment (dizcology) [reply tpu/hptuning/resnet-hypertune/resnet/hypertune_hook.py]: This tensor is retrieved in a session run hook - perhaps there are better ways of getting a specific tensor in a hook?
